clustermesh-apiserver/clustermesh/root.go

```diff
@@ -97,13 +99,53 @@ func registerHooks(lc cell.Lifecycle, params parameters) error {
 				return err
 			}
 
-			startServer(ctx, params.ClusterInfo, params.EnableExternalWorkloads, params.Clientset, backend, params.Resources, params.StoreFactory, params.SyncState)
+			startServer(ctx, params.CrdSyncPromise, params.ClusterInfo, params.EnableExternalWorkloads, params.Clientset, backend, params.Resources, params.StoreFactory, params.SyncState)
 			return nil
 		},
 	})
 	return nil
 }
 
+type syncCRDsPromiseParams struct {
+	cell.In
+
+	Lifecycle cell.Lifecycle
+	Clientset k8sClient.Clientset
+	Resources *synced.Resources
+	APIGroups *synced.APIGroups
+}
+
+func newSyncCRDsPromise(params syncCRDsPromiseParams) promise.Promise[resource.CRDSync] {
+	crdSyncResolver, crdSyncPromise := promise.New[resource.CRDSync]()
+	if !params.Clientset.IsEnabled() || option.Config.DryMode {
+		crdSyncResolver.Reject(fmt.Errorf("disabled"))
+		return crdSyncPromise
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	var wg sync.WaitGroup
+
+	params.Lifecycle.Append(cell.Hook{
+		OnStart: func(cell.HookContext) error {
+			wg.Add(1)
+			go func() {
+				err := synced.SyncCRDs(ctx, params.Clientset, synced.ClusterMeshAPIServerResourceNames(), params.Resources, params.APIGroups)
+				crdSyncResolver.Resolve(err)
+				wg.Done()
+			}()
+			return nil
+		},
+		OnStop: func(cell.HookContext) error {
+			cancel()
+			wg.Wait()
+			return nil
+		},
+	})
+
+	return crdSyncPromise
+}
```

Comment:
I'd suggest to generalize this logic and extract it into a separate cell which just takes the list of resources to wait for as a parameter. So that we don't duplicate the logic in multiple places, which will inevitably start diverging over time.

Reply:
Done, thanks for the suggestion!